packages/common-ts/src/base-service/base-service-v2.ts

```diff
@@ -0,0 +1,232 @@
+import snakeCase from 'lodash/snakeCase'
+
+/* Imports: Internal */
+import { Logger } from '../common/logger'
+import { Metric } from './metrics'
+
+export type Options = {
+  [key: string]: any
+}
+
+export type OptionsSpec<TOptions extends Options> = {
+  [P in keyof Required<TOptions>]: {
+    validator: (spec?: Spec<TOptions[P]>) => ValidatorSpec<TOptions[P]>
+    desc: string
+    default?: TOptions[P]
+  }
+}
+
+export type MetricsV2 = {
+  [key: string]: Metric
+}
+
+export type MetricsSpec<TMetrics extends MetricsV2> = {
+  [P in keyof Required<TMetrics>]: {
+    type: new (configuration: any) => TMetrics[P]
+    desc: string
+    labels?: string[]
+  }
+}
+
+/**
+ * BaseServiceV2 is an advanced but simple base class for long-running TypeScript services.
+ */
+export abstract class BaseServiceV2<
+  TOptions extends Options,
+  TMetrics extends MetricsV2,
+  TServiceState
+> {
+  /**
+   * Whether or not the service will loop.
+   */
+  protected loop: boolean
+
+  /**
+   * Waiting period in ms between loops, if the service will loop.
+   */
+  protected loopIntervalMs: number
+
+  /**
+   * Logger class for this service.
+   */
+  protected logger: Logger
+
+  /**
+   * Service state, persisted between loops.
+   */
+  protected state: TServiceState
+
+  /**
+   * Service options.
+   */
+  protected readonly options: TOptions
+
+  /**
+   * Metrics.
+   */
+  protected readonly metrics: TMetrics
+
+  /**
+   * @param params Options for the construction of the service.
+   * @param params.name Name for the service. This name will determine the prefix used for logging,
+   * metrics, and loading environment variables.
+   * @param params.optionsSpec Settings for input options. You must specify at least a
+   * description for each option.
+   * @param params.metricsSpec Settings that define which metrics are collected. All metrics that
+   * you plan to collect must be defined within this object.
+   * @param params.options Options to pass to the service.
+   * @param params.loops Whether or not the service should loop. Defaults to true.
+   * @param params.loopIntervalMs Loop interval in milliseconds. Defaults to zero.
+   */
+  constructor(params: {
+    name: string
+    optionsSpec: OptionsSpec<TOptions>
+    metricsSpec: MetricsSpec<TMetrics>
+    options?: Partial<TOptions>
+    loop?: boolean
+    loopIntervalMs?: number
+  }) {
+    this.loop = params.loop !== undefined ? params.loop : true
+    this.loopIntervalMs =
+      params.loopIntervalMs !== undefined ? params.loopIntervalMs : 0
+    this.state = {} as TServiceState
+
+    // Use commander as a way to communicate info about the service. We don't actually *use*
+    // commander for anything besides the ability to run `ts-node ./service.ts --help`.
+    const program = new Command()
+    for (const [optionName, optionSpec] of Object.entries(params.optionsSpec)) {
+      program.addOption(
+        new Option(`--${optionName.toLowerCase()}`, `${optionSpec.desc}`).env(
```

Comment:
note that this can be done with bcfg to remove a dep

Reply:
https://github.com/handshake-org/hsd/blob/9c00cff71a3daf06baad2bb2ec23fe9a437a9998/lib/node/node.js#L38
https://github.com/handshake-org/hsd/blob/9c00cff71a3daf06baad2bb2ec23fe9a437a9998/lib/node/node.js#L45

Reply:
I don't think you need to define upfront uning bcfg which config options there are

Reply:
I did it with commander mainly so that I could get the `--help` command by default (you can run `ts-node ./path/to/service --help` and get useful help info).